### ISSUE TITLE
IP and policy removal not working on connection update

### ISSUE THREAD

**Issue by LionelJouin** [networkservicemesh/sdk-kernel#426]
After some testing with this change: https://github.com/networkservicemesh/sdk/issues/1231. I found it was possible to add new IP addresses and policy routes by just requesting (updating) a new time without having to close the connection. 

The issue I noticed is that it is not possible to remove any IP or policy route while they are removed from the IP context in the request.

**Comment by denis-tingaikin** [comment]
I feel this is totally related to https://github.com/networkservicemesh/cmd-forwarder-vpp/issues/441

To resolve the issue we should rework each chain element from sdk-vpp and check do we have a change? If so then we need to delete previosly created things.

**Comment by denis-tingaikin** [comment]
cc @edwarnicke

**Comment by LionelJouin** [comment]
Right, it is related. Just in my use case, the NSC is changing intentionally the content of the IP context. From my tests, the ipadress and iprule chain elements are working fine for adding but not for removing.

**Comment by edwarnicke** [comment]
@LionelJouin That likely is related to needing to update the chain elements that do that work.

@denis-tingaikin One other thing to watch carefully is that we are only doing the work on the return stroke of the call in those chain elements.  If the NSE is OK allowing the NSC to update IP context... that's fine... but its not OK to act on NSC requests to update IPContext until the NSE has *returned* those updated (meaning it thinks they are valid).

@LionelJouin Does that make sense to you?

**Comment by LionelJouin** [comment]
yes, it's clear to me.

**Comment by denis-tingaikin** [comment]
@LionelJouin https://github.com/networkservicemesh/sdk/pull/1234 is merged. Could you test is it solves the problem?

**Comment by LionelJouin** [comment]
I tried it, it works fine, It solved this issue: https://github.com/networkservicemesh/sdk/issues/1231

**Comment by edwarnicke** [comment]
@LionelJouin and does it also solve your issue here with IP policy removal on connection update?  Or is that still open?

**Comment by LionelJouin** [comment]
no, this issue is still open (https://github.com/networkservicemesh/sdk-kernel/issues/426).

**Comment by LionelJouin** [comment]
@edwarnicke @denis-tingaikin How the forwarder resiliency could be handled? Handling IP addresses would be easy since it is possible to get the IP addresses by checking the interface directly and then compare them to the ones in the IPContext to find which ones have to be added or removed. But I believe this would be more complex for the policy routes, how the forwarder could remember which rule/table belongs to which policy? (in case the forwarder crashes/restarts/...)

**Comment by glazychev-art** [comment]
Hello @LionelJouin,
Maybe I misunderstood you. Is the problem still present after this fix? - https://github.com/networkservicemesh/sdk-kernel/pull/437

**Comment by LionelJouin** [comment]
Hi, yes, I think so. For instance, if the forwarder restarts, then the [policies map](https://github.com/glazychev-art/sdk-kernel/blob/3591e4991ec102431e4c783500148c4cb0507a04/pkg/kernel/networkservice/connectioncontextkernel/ipcontext/iprule/gen.go#L27) will be emptied, and then the forwarder will loose track of which ip tables/rules it was managing previously.

**Comment by edwarnicke** [comment]
@LionelJouin One other thing to keep in mind is that if your NSE wants to *trigger* a change... it should probably use:

https://github.com/networkservicemesh/api/blob/main/pkg/api/networkservice/connection.proto#L36

Which is to say:
1. NSE sends a ConnectionEvent over Monitor towards the NSE with State set to REFRESH_REQUESTED
2. This will cause the NSC to send a new Request
3. The NSE can return a changed Connection.ConnectionContext

Basically... this way you don't have to wait around for a Refresh to happen at Expiration.

**Comment by glazychev-art** [comment]
@LionelJouin 
As far as I can see, all main PRs have been merged.
Can we close this issue?

**Comment by LionelJouin** [comment]
There is still 2e2 tests to develop and other properties to support. We could keep this issue, otherwise I would open new ones for the e2e and properties

**Comment by edwarnicke** [comment]
@LionelJouin Agree on tracking the other issues.  I'd suggest opening new issues and referencing this one.
